File: ios/StockbookTests/MoneyTests.swift
import Testing
@testable import Stockbook

@Suite("Currency formatting")
struct MoneyTests {

    @Test("Integers render without decimals")
    func integers() {
        #expect(Money.text(194) == "SAR 194")
        #expect(Money.text(0) == "SAR 0")
    }

    @Test("Non-integers render to exactly two places")
    func fractions() {
        #expect(Money.text(0.25) == "SAR 0.25")
        #expect(Money.text(0.5) == "SAR 0.50")
    }

    @Test("Thousands are grouped en-US")
    func grouping() {
        #expect(Money.text(1240) == "SAR 1,240")
        #expect(Money.text(1240.5) == "SAR 1,240.50")
    }

    @Test("Rounding happens before the integer test")
    func rounding() {
        // 193.999 rounds to 194, which is then an integer and loses its decimals.
        #expect(Money.text(193.999) == "SAR 194")
        #expect(Money.text(0.005) == "SAR 0.01")
    }

    @Test("Negative zero does not print a minus sign")
    func negativeZero() {
        #expect(Money.text(-0.0) == "SAR 0")
    }

    @Test("A signed figure carries the minus in front of the symbol")
    func signedFigures() {
        // `SAR -150` is what putting the symbol first gets you, and it is not
        // how anybody writes a negative.
        #expect(Money.signed(-150) == "-SAR 150")
        #expect(Money.signed(150) == "SAR 150")
        #expect(Money.signed(-12.5, in: .inr) == "-₹12.50")
    }

    @Test("A signed figure that rounds to nothing carries no sign")
    func signedZero() {
        // A loss of nothing is not a loss.
        #expect(Money.signed(-0.004) == "SAR 0")
        #expect(Money.signed(-0.0) == "SAR 0")
    }

    @Test("The symbol is configurable")
    func symbol() {
        #expect(Money.text(12, in: .inr) == "₹12")
    }

    @Test("Parsing tells empty from zero")
    func parsing() {
        #expect(Money.parse("") == nil)
        #expect(Money.parse("   ") == nil)
        #expect(Money.parse("0") == 0)
        #expect(Money.parse("1,240") == 1240)
        #expect(Money.parse("abc") == nil)
    }
}

@Suite("Copy helpers")
struct CopyTests {

    @Test("First name is the first whitespace-separated word")
    func firstName() {
        #expect("Ahmed Al-Amri".firstName == "Ahmed")
        #expect("  Khalid  ".firstName == "Khalid")
        #expect("".firstName == "")
    }
}

File: android/core/src/test/kotlin/com/stockbook/core/MoneyTests.kt
package com.stockbook.core

import com.stockbook.core.model.Currency
import com.stockbook.core.money.Money
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull

/** The exact expectations the iOS build asserts, so the two never diverge. */
class MoneyTests {

    @Test
    fun `whole numbers carry no decimals`() {
        assertEquals("SAR 194", Money.text(194.0))
        assertEquals("SAR 0", Money.text(0.0))
    }

    @Test
    fun `anything else carries exactly two`() {
        assertEquals("SAR 0.25", Money.text(0.25))
        assertEquals("SAR 0.50", Money.text(0.5))
    }

    @Test
    fun `thousands are grouped`() {
        assertEquals("SAR 1,240", Money.text(1240.0))
        assertEquals("SAR 1,240.50", Money.text(1240.5))
    }

    @Test
    fun `values round to the currency's minor unit`() {
        assertEquals("SAR 194", Money.text(193.999))
        assertEquals("SAR 0.01", Money.text(0.005))
    }

    @Test
    fun `negative zero prints as zero`() {
        assertEquals("SAR 0", Money.text(-0.0))
    }

    @Test
    fun `the symbol is the currency's`() {
        assertEquals("₹12", Money.text(12.0, Currency.INR))
    }

    @Test
    fun `parsing tells empty from zero`() {
        assertEquals(0.0, Money.parse("0"))
        assertNull(Money.parse(""))
        assertNull(Money.parse("   "))
        assertNull(Money.parse("nonsense"))
        assertEquals(1240.5, Money.parse("1,240.50"))
    }
}
